### DISCUSSION TITLE
Renaming pool/ROOT/mydist and boot.

### DISCUSSION THREAD

**Discussion by dandudikof** [zbm-dev/zfsbootmenu#517] [Questions]
Ran in to an issue of where i made a copy of pool/ROOT/mydist1@snapshot
with (zfs send | zfs recv) to pool/ROOT/mydist2 and continued to work in mydist1 
without realizing it (wanted to be in mydist2). 

So after the fact i renamed mydist1 <-> mydist2 from a zfs rescue boot stick,
but up on reboot with zfsbootmenu mydist2 would boot with root of mydist1 
and vice versa for dist1 booting with root of mydist2. 

Fixed by chrooting from zfs rescue and ran update-initramfs -ck all in each.

Just wanted to clear up this behavior. Does the root of the system 
get hard coded in to initramfs ? 

Unlike grub where it is passed as kernel boot parameter from grub.cfg?

thanx 

dan

**Comment by ahesford** [comment]
ZFSBootMenu imports all pools and checks the `bootfs` property of the first pool it finds (in an arbitrary order, which may be alphabetical) to determine what to boot by default. When it boots a kernel in that environment, or you select another environment explicitly, it passes a root argument that refers to what will boot.

The root argument can take different forms depending on the distribution, so we make an effort to figure out what distribution will boot and choose the proper form. We do this by trying to read `os-release`. If this fails or we don't recognize a particular distribution, we pick a fallback that should work on Void Linux and several other distributions (especially those that use Dracut for initramfs images.)

This behavior can be overridden if necessary, but because your environments were booting before, I doubt they were receiving the wrong root argument. More likely, something about your environments is overriding whatever you expect to happen.

You can see what ZBM has passed to your boot environment by inspecting the contents of `/proc/cmdline`.

**Reply by dandudikof** [reply]
Running devuan 4/5 . with non dracut initramfs. 

Where would root parameter be sourced from in this case ?

**Reply by ahesford** [reply]
Does Devuan not include an `/etc/os-release` or `/usr/lib/os-release`? ZBM reads those files from the filesystem, not an initramfs image.

**Reply by dandudikof** [reply]
Still not clear of how root is set.

And renaming a rpool/ROOT/dist1 to rpool/ROOT/dist2
and creating a new rpool/ROOT/dist1 
while bootfs=rpool/ROOT/dist1 was untouched
rebooting now default to rpool/ROOT/dist2

So does zfsbootmenu keep track with uuids ?


#cat /etc/os-release

PRETTY_NAME="Devuan GNU/Linux 5 (daedalus)"
NAME="Devuan GNU/Linux"
VERSION_ID="5"
VERSION="5 (daedalus)"
VERSION_CODENAME="daedalus"
ID=devuan
ID_LIKE=debian
HOME_URL="https://www.devuan.org/"
SUPPORT_URL="https://devuan.org/os/community"
BUG_REPORT_URL="https://bugs.devuan.org/"

**Reply by ahesford** [reply]
ZFS itself ought to change the value of the `bootfs` property when you rename a filesystem. ZFSBootMenu doesn't keep track of anything; it reads the value of the property and will try to boot that unless you interrupt the countdown and choose something else.

**Comment by zdykstra** [comment]
As an aside, there's no need to use a rescue CD. There's a full shell built into ZFSBootMenu, along with the ability to chroot into any boot environment. Hit MOD+W on the main screen and then you can drop to a shell via MOD+R, or enter into a chroot of the selected environment via MOD+J.

**Reply by dandudikof** [reply]
I am aware that this functionality exists. 

Just have my own usb boot sticks with custom install and my own scrips/helpers
for various occasions and hardware types, and (uefi/bios on gpt)(bios on mbr)
etc,,, for older computers and funky uefi implementations.
